stats/opentelemetry/internal/csm/pluginoption.go

```diff
@@ -0,0 +1,341 @@
+	// The value of “csm.remote_workload_canonical_service” comes from
+	// MetadataExchange with the key “canonical_service”. (Note that this should
+	// be read even if the remote type is unknown.)
+	appendToLabelsFromMetadata(labels, "csm.remote_workload_canonical_service", "canonical_service", fields)
+
+	// Unset/unknown types, and types that aren't GKE or GCP return early with
+	// just local labels, remote_workload_type and
+	// remote_workload_canonical_service labels.
+	typeVal := labels["csm.remote_workload_type"]
+	if typeVal != "gcp_kubernetes_engine" && typeVal != "gcp_compute_engine" {
+		return labels
+	}
+	// GKE and GCE labels.
+	appendToLabelsFromMetadata(labels, "csm.remote_workload_project_id", "project_id", fields)
+	appendToLabelsFromMetadata(labels, "csm.remote_workload_location", "location", fields)
+	appendToLabelsFromMetadata(labels, "csm.remote_workload_name", "workload_name", fields)
+	if typeVal == "gcp_compute_engine" {
+		return labels
+	}
+
+	// GKE only labels.
+	appendToLabelsFromMetadata(labels, "csm.remote_workload_cluster_name", "cluster_name", fields)
+	appendToLabelsFromMetadata(labels, "csm.remote_workload_namespace_name", "namespace_name", fields)
+	return labels
+}
+
+// appendToLabelsFromMetadata appends to the labels map passed in. It sets
+// "unknown" if the metadata is not found in the struct proto, or if the value
+// is not a string value.
+func appendToLabelsFromMetadata(labels map[string]string, labelKey string, metadataKey string, metadata map[string]*structpb.Value) {
+	labelVal := "unknown"
+	if metadata != nil {
+		if metadataVal, ok := metadata[metadataKey]; ok {
+			if _, ok := metadataVal.GetKind().(*structpb.Value_StringValue); ok {
+				labelVal = metadataVal.GetStringValue()
+			}
+		}
+	}
+	labels[labelKey] = labelVal
+}
+
+// appendToLabelsFromResource appends to the labels map passed in. It sets
+// "unknown" if the resourceKey is not found in the attribute set or is not a
+// string value, the string value otherwise.
+func appendToLabelsFromResource(labels map[string]string, labelKey string, resourceKey attribute.Key, set *attribute.Set) {
+	labelVal := "unknown"
+	if set != nil {
+		if resourceVal, ok := set.Value(resourceKey); ok && resourceVal.Type() == attribute.STRING {
+			labelVal = resourceVal.AsString()
+		}
+	}
+	labels[labelKey] = labelVal
+}
+
+// appendToLabelsFromEnv appends an environment variable key value pair to the
+// labels passed in. It sets "unknown" if environment variable is unset, the
+// environment variable otherwise.
+func appendToLabelsFromEnv(labels map[string]string, labelKey string, envvar string) {
+	envvarVal := "unknown"
+	if val, ok := os.LookupEnv(envvar); ok {
+		envvarVal = val
+	}
+	labels[labelKey] = envvarVal
+}
+
+var (
+	// This function will be overridden in unit tests.
+	getAttrSetFromResourceDetector = func() *attribute.Set {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```

Comment:
Only main may call `context.Background`...  Seems like the thing that creates the plugin needs to accept a context parameter and we plumb it here?

Reply:
Hmm ok I guess our global csm plugin layer will also take a context in it's constructor.

Reply:
Added context flow through NewPluginOption() -> getAttrSetFromResourceDetector().